core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1366,6 +1376,17 @@ class Partition(val topicPartition: TopicPartition,
           fetchParams.replicaId,
           fetchPartitionData
         )
+
+        // Fence the fetch request with stale broker epoch from a rebooted follower.
+        if (metadataCache.isInstanceOf[KRaftMetadataCache]) {
+          val brokerEpoch = fetchParams.replicaEpoch
+          val currentBrokerEpoch = replica.stateSnapshot.brokerEpoch.getOrElse(-1L)
+          if (brokerEpoch != -1 && brokerEpoch < currentBrokerEpoch) {
+            throw new StaleBrokerEpochException(s"Received fetch request for $topicPartition with stale broker " +
+              s"epoch=$brokerEpoch. The expected broker epoch= $currentBrokerEpoch.")
+          }
+        }
```

Comment:
We would also need to update the `Replica.updateFetchState` call to check the brokerEpoch before trying to apply a fetch update. Otherwise we can pass this check and race for an update.

I don't think we need to check whether we are in KRaft mode or not. The broker epoch should be monotonic in both zookeeper and kraft mode.

I also don't think we can throw `StaleBrokerEpochException` since this would introduce a new error code returned on fetch responses without a KIP.

Reply:
Make sense. Then just abort the fetch state update.